dns.go

```diff
@@ -20,29 +20,32 @@ import (
 // OpportunisticResolver opportunistically tries encrypted DNS over TLS
 // using the local resolver.
 var OpportunisticResolver = &net.Resolver{
-	Dial:     opportunisticDial,
+	Dial:     getAnOpportunisticDialerFromDialFunc(nil),
 	PreferGo: true,
 }
 
-func opportunisticDial(ctx context.Context, network, address string) (net.Conn, error) {
-	host, port, _ := net.SplitHostPort(address)
-	if (port == "53" || port == "domain") && notBadServer(address) {
-		deadline, ok := ctx.Deadline()
-		if ok && deadline.After(time.Now().Add(2*time.Second)) {
-			var d net.Dialer
-			d.Timeout = time.Second
-			tlsAddr := net.JoinHostPort(host, "853")
-			tlsConf := tls.Config{InsecureSkipVerify: true}
-			conn, _ := tls.DialWithDialer(&d, "tcp", tlsAddr, &tlsConf)
-			if conn != nil {
-				return conn, nil
+func getAnOpportunisticDialerFromDialFunc(dial DialFunc) DialFunc {
+	if dial == nil {
+		var d net.Dialer
+		dial = d.DialContext
+	}
+	return func(ctx context.Context, network, address string) (net.Conn, error) {
+		host, port, _ := net.SplitHostPort(address)
+		if (port == "53" || port == "domain") && notBadServer(address) {
+			deadline, ok := ctx.Deadline()
+			if ok && deadline.After(time.Now().Add(2*time.Second)) {
+				tlsAddr := net.JoinHostPort(host, "853")
+				rawConn, _ := dial(ctx, "tcp", tlsAddr)
+				if rawConn != nil {
+					tlsConf := tls.Config{InsecureSkipVerify: true}
+					return tls.Client(rawConn, &tlsConf), nil
```

Comment:
This is also different from above. This version doesn't complete the handshake here, it will only do so lazily on the first `Write`. By then, if TLS config is deemed unacceptable, fallback is impossible.